lib/connextclient/ConnextClient.ts

```diff
@@ -0,0 +1,699 @@
+    let tokenAddress: string;
+    let lockTimeout: number | undefined;
+    try {
+      let secret;
+      if (deal.role === SwapRole.Maker) {
+        // we are the maker paying the taker
+        amount = deal.takerUnits;
+        tokenAddress = this.tokenAddresses.get(deal.takerCurrency)!;
+        const executeTransfer = this.executeHashLockTransfer({
+          assetId: tokenAddress,
+          amount: `${amount}`,
+          timelock: deal.takerCltvDelta.toString(),
+          lockHash: `0x${deal.rHash}`,
+          recipient: deal.destination!,
+        });
+        // @ts-ignore
+        const [executeTransferResponse, preimage] = await Promise.all([
+          executeTransfer,
+          waitForPreimageByHash(this, deal.rHash),
+        ]);
+        secret = preimage;
+      } else {
+        // we are the taker paying the maker
+        amount = deal.makerUnits;
+        tokenAddress = this.tokenAddresses.get(deal.makerCurrency)!;
+        lockTimeout = deal.makerCltvDelta!;
+        secret = deal.rPreimage!;
+        const executeTransfer = this.executeHashLockTransfer({
+          assetId: tokenAddress,
+          amount: `${amount}`,
+          timelock: lockTimeout.toString(),
+          lockHash: `0x${deal.rHash}`,
+          recipient: deal.destination!,
+        });
+        await executeTransfer;
+      }
+      return secret;
+    } catch (err) {
+      switch (err.code) {
+        case 'ECONNRESET':
+        case errorCodes.UNEXPECTED:
+        case errorCodes.TIMEOUT:
+        case errorCodes.SERVER_ERROR:
+        case errorCodes.INVALID_TOKEN_PAYMENT_RESPONSE:
+          throw swapErrors.UNKNOWN_PAYMENT_ERROR(err.message);
+        default:
+          throw swapErrors.FINAL_PAYMENT_ERROR(err.message);
+      }
+    }
+  }
+
+  public addInvoice = async (
+    { rHash: expectedHash, units: expectedUnits, expiry: expectedTimelock, currency: expectedCurrency }:
+    { rHash: string, units: number, expiry?: number, currency?: string },
+  ) => {
+    if (!expectedCurrency) {
+      throw new Error('cannot add invoice, expected currency is missing');
+    }
+    if (!expectedTimelock) {
+      throw new Error('cannot add invoice, expiry is missing');
+    }
+    const expectedTokenAddress = this.getTokenAddress(expectedCurrency);
+    const expectedIncomingTransfer: ExpectedIncomingTransfer = {
+      rHash: expectedHash,
+      units: expectedUnits,
+      expiry: expectedTimelock,
+      tokenAddress: expectedTokenAddress,
+    };
+    // TODO: should we also store this in the database?
+    this.expectedIncomingTransfers.set(expectedHash, expectedIncomingTransfer);
+  }
+
+  /**
+   * Resolve a HashLock Transfer on the Connext network.
+   */
+  public settleInvoice = async (_rHash: string, rPreimage: string) => {
+    await this.sendRequest('/hashlock-resolve', 'POST', {
+      lockHash: `0x${rPreimage}`,
+    });
+  }
+
+  public removeInvoice = async () => {
+    // not implemented, connext does not use invoices
```

Comment:
Should we remove from the expected incoming transfers here?

Reply:
Good catch!